In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.font_manager as fm
import matplotlib
import timeit
import feather
%matplotlib inline

In [2]:
res = pd.read_feather('DBus_data_first_clean_ian.feather')

In [5]:
res.shape

(70904023, 11)

In [4]:
# Attaching leading zeroes to VehicleJourneyIDs and StopIDs of less than 4 digits
# code from: https://stackoverflow.com/questions/33243763/pandas-add-leading-0-to-string-values-so-all-values-are-equal-len

res.VehicleJourneyID = res.VehicleJourneyID.apply(lambda x: str(int(x)).zfill(4))
res.StopID = res.StopID.apply(lambda x: str(int(x)).zfill(4))

In [6]:
res.head(4)

Timestamp JourneyPatternID  TimeFrame VehicleJourneyID       Lon  \
0 2013-01-31 00:00:01         00131005 2013-01-30             3406 -6.273923   
1 2013-01-31 00:00:03         00071003 2013-01-30             6069 -6.230700   
2 2013-01-31 00:00:03         00671001 2013-01-30             0905 -6.325533   
3 2013-01-31 00:00:03         00831001 2013-01-30             5613 -6.284449   

         Lat  Delay VehicleID StopID  AtStop        Day  
0  53.343307   -235     33608   1998   False  Wednesday  
1  53.317432      0     43003   0008    True  Wednesday  
2  53.346302   -396     33438   2248   False  Wednesday  
3  53.420197   -490     40012   1552   False  Wednesday

## ------------------------Start: The same as last version ---------------------------- ##

### Part I. Second stage cleaning

#### 1. Delete those rows of same TimeFrame + JourneyPatternID  +  VehicleJourneyID +  VehicleID happen under n times
0711: test n = 10

In [7]:
gb = res.groupby(['TimeFrame', 'JourneyPatternID', 'VehicleJourneyID', 'VehicleID'])

In [8]:
# Delete those count is less than 10

gbcount = gb['AtStop'].count()
idx = pd.DataFrame(gbcount[gbcount < 10])

In [9]:
# Set the index back to data

idx.reset_index(['TimeFrame', 'JourneyPatternID', 'VehicleJourneyID', 'VehicleID'], inplace=True)

In [10]:
# Merge back to res to see which one should be keep

merge = pd.merge(res, idx, on=['TimeFrame', 'JourneyPatternID', 'VehicleJourneyID', 'VehicleID'], how='outer',suffixes='oa')

# If null means that row should keep

res = merge[merge['AtStopa'].isnull()]

# Clean up

res.drop('AtStopa', axis=1, inplace=True)
res = res.rename(columns = {'AtStopo':'AtStop'})
del merge

/Users/EmmaChang/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()


#### 2. Delete those rows of same TimeFrame + VehicleJourneyID  but exist over one JourneyPatternID

In [11]:
# Due to make following function work need to change datatype here

res["JourneyPatternID"] = res["JourneyPatternID"].astype("object")

# Check how many unique JourneyPatternID under that group (if over 1 is abnormal)

gb = res.groupby(['TimeFrame', 'VehicleJourneyID'])
idx = pd.DataFrame(gb['JourneyPatternID'].unique())
idx['JourneyPatternID'] = idx['JourneyPatternID'].apply(lambda x: len(x))

In [12]:
# Get those rows which unique JourneyPatternID are over 1

idx = idx[idx.JourneyPatternID > 1]

# Set the index back to data

idx.reset_index(['TimeFrame', 'VehicleJourneyID'], inplace=True)


In [13]:
# Merge with res then can see if is null at the column JourneyPatternIDA then should keep
# suffixes='OA' : is set column name to JourneyPatternIDO and JourneyPatternIDA

merge = pd.merge(res, idx, on=['TimeFrame', 'VehicleJourneyID'], how='outer',suffixes='OA')

# Keep those JourneyPatternIDA is null

res = merge[merge['JourneyPatternIDA'].isnull()]

# Clean up

res.drop('JourneyPatternIDA', axis=1, inplace=True)
res = res.rename(columns = {'JourneyPatternIDO':'JourneyPatternID'})
del merge

/Users/EmmaChang/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [15]:
# Store as feather

res = res.reset_index()
res.to_feather('DBus_stage_three_clean_v1_step1.feather')

In [78]:
res.shape

(70555585, 12)

## ------------------------End: The same as last version ---------------------------- ##

### Part II. Delete stopID not in that trip_id (same meaning as JourneyPatternID)

In [2]:
res = pd.read_feather('DBus_stage_three_clean_v1_step1.feather')
res.shape

(70555585, 12)

#### Step 1. Prepare datafram to compare

#### 1. Get the dataframe that has shape_id, route, sequence of stops and stopID base on the "googletransit20130315-1546/stop_times.csv" data

In [3]:
# Read in stop_times.csv and trips.csv
stop_times_2012 = pd.read_csv('stop_times.csv')
trips_2012 = pd.read_csv('trips.csv')

# Merge by trip_id
# trip_id here has the same meaning of JourneyPatternID
merge = pd.merge(stop_times_2012, trips_2012, on='trip_id', how='outer')

# Keep only necessary columns
merge.drop(['arrival_time','departure_time','pickup_type','drop_off_type','service_id','shape_dist_traveled'], axis=1, inplace=True)

In [4]:
# Seporate trip_id to route, stop_id to StopID
# Route
merge['route_short'] = merge['trip_id'].apply(lambda x: x[x.index('-')+1:])
merge['route_short'] = merge['route_short'].apply(lambda x: x[: x.index('-')])
merge['route_short'] = merge['route_short'].apply(lambda x: str(x).zfill(4))

# StopID
merge['StopID'] = merge['stop_id'].apply(lambda x: x[-4:])

In [5]:
# Find StopID and the sequence of that shape_id
gb = merge.groupby(['shape_id', 'route_short', 'direction_id','stop_sequence', 'StopID'])
gbc = gb.count()
gbc.reset_index(['shape_id', 'route_short', 'direction_id','stop_sequence', 'StopID'], inplace=True)
transit_shapeID_stopID = gbc.drop(['trip_id','stop_id','route_id','trip_headsign'], axis=1)

In [6]:
# Now we have trip_id to its corresponding StopID and sequence
transit_shapeID_stopID

shape_id route_short  direction_id  stop_sequence StopID
0       0-1-y12-1.1.O        0001             0              1   0226
1       0-1-y12-1.1.O        0001             0              2   0228
2       0-1-y12-1.1.O        0001             0              3   0229
3       0-1-y12-1.1.O        0001             0              4   0227
4       0-1-y12-1.1.O        0001             0              5   0230
5       0-1-y12-1.1.O        0001             0              6   0231
6       0-1-y12-1.1.O        0001             0              7   1641
7       0-1-y12-1.1.O        0001             0              8   1642
8       0-1-y12-1.1.O        0001             0              9   0213
9       0-1-y12-1.1.O        0001             0             10   0214
10      0-1-y12-1.1.O        0001             0             11   4432
11      0-1-y12-1.1.O        0001             0             12   0119
12      0-1-y12-1.1.O        0001             0             13   0044
13      0-1-y12-1.1.O        0001             0             14   0045
14      0-1-y12-1.1.O        0001             0             15   0046
15      0-1-y12-1.1.O        0001             0             16   0047
16      0-1-y12-1.1.O        0001             0             17   0048
17      0-1-y12-1.1.O        0001             0             18   0049
18      0-1-y12-1.1.O        0001             0             19   0050
19      0-1-y12-1.1.O        0001             0             20   0051
20      0-1-y12-1.1.O        0001             0             21   0052
21      0-1-y12-1.1.O        0001             0             22   0265
22      0-1-y12-1.1.O        0001             0             23   0271
23      0-1-y12-1.1.O        0001             0             24   0340
24      0-1-y12-1.1.O        0001             0             25   0350
25      0-1-y12-1.1.O        0001             0             26   0351
26      0-1-y12-1.1.O        0001             0             27   0352
27      0-1-y12-1.1.O        0001             0             28   0353
28      0-1-y12-1.1.O        0001             0             29   0354
29      0-1-y12-1.1.O        0001             0             30   0355
...               ...         ...           ...            ...    ...
23296  0-9-y12-1.13.I        0009             1              9   2431
23297  0-9-y12-1.13.I        0009             1             10   2432
23298  0-9-y12-1.13.I        0009             1             11   2433
23299  0-9-y12-1.13.I        0009             1             12   2434
23300  0-9-y12-1.13.I        0009             1             13   2435
23301  0-9-y12-1.13.I        0009             1             14   2436
23302  0-9-y12-1.13.I        0009             1             15   2437
23303  0-9-y12-1.13.I        0009             1             16   2438
23304  0-9-y12-1.13.I        0009             1             17   2439
23305  0-9-y12-1.13.I        0009             1             18   2440
23306  0-9-y12-1.13.I        0009             1             19   2441
23307  0-9-y12-1.13.I        0009             1             20   2442
23308  0-9-y12-1.13.I        0009             1             21   2443
23309  0-9-y12-1.13.I        0009             1             22   2444
23310  0-9-y12-1.13.I        0009             1             23   1344
23311  0-9-y12-1.13.I        0009             1             24   1345
23312  0-9-y12-1.13.I        0009             1             25   1347
23313  0-9-y12-1.13.I        0009             1             26   1348
23314  0-9-y12-1.13.I        0009             1             27   1349
23315  0-9-y12-1.13.I        0009             1             28   1350
23316  0-9-y12-1.13.I        0009             1             29   1348
23317  0-9-y12-1.13.I        0009             1             30   1352
23318  0-9-y12-1.13.I        0009             1             31   1353
23319  0-9-y12-1.13.I        0009             1             32   1354
23320  0-9-y12-1.13.I        0009             1             33   1355
23321  0-9-y12-1.13.I  

#### 2. Get the dataframe that has JourneyPatternID to its possible StopID from the data clean from second stage

In [7]:
# Get JourneyPatternID and its StopID
temp = res[res.AtStop == True]
gb = temp.groupby(['JourneyPatternID', 'StopID'])
gc = gb.count()

In [8]:
# Clean up
gbc = gc[~gc.Timestamp.isnull()] # Doing this here is cause the feather will include all column, so clean up a little here
gbc.reset_index(['JourneyPatternID', 'StopID'], inplace=True)
gps_JPID_StopID = gbc.loc[:,['JourneyPatternID', 'StopID', 'AtStop']]
gps_JPID_StopID = gps_JPID_StopID.rename(columns = {'AtStop':'Count'})

In [9]:
# Separate JourneyPatternID to route and direction
gps_JPID_StopID['route_short'] = gps_JPID_StopID['JourneyPatternID'].str[:4]
gps_JPID_StopID['Direction'] = gps_JPID_StopID['JourneyPatternID'].str[4:]


# Now we have JourneyPatternID to its possible StopID
gps_JPID_StopID

JourneyPatternID StopID    Count route_short Direction
0             00010001   0119    688.0        0001      0001
1             00010001   1641    352.0        0001      0001
2             00010001   1642    648.0        0001      0001
3             00010001   0213   1734.0        0001      0001
4             00010001   0214    266.0        0001      0001
5             00010001   0226  23931.0        0001      0001
6             00010001   0227    716.0        0001      0001
7             00010001   0228    203.0        0001      0001
8             00010001   0229    689.0        0001      0001
9             00010001   0230    258.0        0001      0001
10            00010001   0231   1370.0        0001      0001
11            00010001   0265   4401.0        0001      0001
12            00010001   0271   9518.0        0001      0001
13            00010001   2804    480.0        0001      0001
14            00010001   0340   4438.0        0001      0001
15            00010001   0350    986.0        0001      0001
16            00010001   0351   1329.0        0001      0001
17            00010001   0352   1523.0        0001      0001
18            00010001   0353   1245.0        0001      0001
19            00010001   0354   2478.0        0001      0001
20            00010001   0355   1608.0        0001      0001
21            00010001   0356   1886.0        0001      0001
22            00010001   0357   1534.0        0001      0001
23            00010001   0372    750.0        0001      0001
24            00010001   0373    303.0        0001      0001
25            00010001   0374   1951.0        0001      0001
26            00010001   0375    180.0        0001      0001
27            00010001   0376    160.0        0001      0001
28            00010001   0377    601.0        0001      0001
29            00010001   0378   1791.0        0001      0001
...                ...    ...      ...         ...       ...
22292             86-2   4396      1.0        86-2          
22293             86-2   5032      7.0        86-2          
22294             86-2   7361      1.0        86-2          
22295         OL77X101   1358    142.0        OL77      X101
22296         OL77X101   1424     31.0        OL77      X101
22297         OL77X101   1436      3.0        OL77      X101
22298         OL77X101   2187      1.0        OL77      X101
22299         OL77X101   2190    157.0        OL77      X101
22300         OL77X101   2371     56.0        OL77      X101
22301         OL77X101   2373     23.0        OL77      X101
22302         OL77X101   2425     21.0        OL77      X101
22303         OL77X101   2427      6.0        OL77      X101
22304         OL77X101   2430     25.0        OL77      X101
22305         OL77X101   2538      7.0        OL77      X101
22306         OL77X101   2607      3.0        OL77      X101
22307         OL77X101   0350   1782.0        OL77      X101
22308         OL77X101   4714   4505.0        OL77      X101
22309         OL77X101   0493   2182.0        OL77      X101
22310         OL77X101   0779     92.0        OL77      X101
22311         PP071001   0279      3.0        PP07      1001
22312         PP071001   3076      1.0        PP07      1001
22313         PP071001   3080      1.0        PP07      1001
22314         PP071001   3222     15.0        PP07      1001
22315         PP071001   3224      1.0        PP07      1001
22316         PP071001   3240      1.0        PP07      1001
22317         PP071001   4705      1.0        PP07      1001
22318         PP071001   0476      8.0        PP07      1001
22319         PP071001   0487      1.0        PP07      1001
22320         PP071001   0493      1.0        PP07      1001
22321         PP071001   4962    180.0        PP07      1001

[22322 rows x 5 columns]

0720 add
### Check if the number of trip_id is match with JourneyPatternID
In this section, we try to find out if the number of trip_id under each route in googletransit2012 is the same number of JourneyPatternID in gps data. For example, for route 007B has trip_id "0-7B-y12-1.176.O" and "0-7B-y12-1.177.I"; JourneyPatternID "007B0001" and "007B1001". So we know at least there is one-to-one mapping. Other scenario see document "compare_route_googletransit_vs_gps.xlsx"

In [10]:
gb_tss = transit_shapeID_stopID.groupby(['route_short','shape_id'])
gb_tss_c = gb_tss.count()
gb_tss_c = gb_tss_c.reset_index(['route_short','shape_id'])
gb_tss_c.drop(['direction_id', 'stop_sequence', 'StopID'], axis=1, inplace=True)
gb_tss_c

route_short           shape_id
0          0001      0-1-y12-1.1.O
1          0001      0-1-y12-1.2.O
2          0001      0-1-y12-1.3.O
3          0001      0-1-y12-1.4.I
4          0001      0-1-y12-1.5.I
5          0004      0-4-y12-1.6.O
6          0004      0-4-y12-1.7.O
7          0004      0-4-y12-1.8.I
8          0004      0-4-y12-1.9.I
9          0007    0-7-y12-1.167.O
10         0007    0-7-y12-1.168.O
11         0007    0-7-y12-1.169.I
12         0007    0-7-y12-1.170.I
13         0007    0-7-y12-1.171.O
14         0007    0-7-y12-1.172.O
15         0007    0-7-y12-1.173.O
16         0007    0-7-y12-1.174.I
17         0007    0-7-y12-1.175.I
18         0008    0-8-y12-1.181.O
19         0008    0-8-y12-1.182.I
20         0008    0-8-y12-1.183.I
21         0009     0-9-y12-1.10.O
22         0009     0-9-y12-1.11.O
23         0009     0-9-y12-1.12.I
24         0009     0-9-y12-1.13.I
25         0011    0-11-y12-1.14.O
26         0011    0-11-y12-1.15.O
27         0011    0-11-y12-1.16.I
28         0011    0-11-y12-1.17.I
29         0013    0-13-y12-1.18.O
..          ...                ...
440        066X  0-66X-y12-1.367.I
441        067X  0-67X-y12-1.370.O
442        067X  0-67X-y12-1.371.O
443        067X  0-67X-y12-1.372.O
444        067X  0-67X-y12-1.373.I
445        067X  0-67X-y12-1.374.I
446        068A  0-68A-y12-1.384.O
447        068A  0-68A-y12-1.385.I
448        069X  0-69X-y12-1.389.O
449        069X  0-69X-y12-1.390.I
450        0747  0-747-y12-1.165.O
451        0747  0-747-y12-1.166.I
452        076A  0-76A-y12-1.393.O
453        076A  0-76A-y12-1.394.I
454        077A  0-77A-y12-1.235.O
455        077A  0-77A-y12-1.236.I
456        079A  0-79A-y12-1.397.O
457        079A  0-79A-y12-1.398.I
458        083A  0-83A-y12-1.407.O
459        083A  0-83A-y12-1.408.O
460        083A  0-83A-y12-1.409.I
461        083A  0-83A-y12-1.410.I
462        084A  0-84A-y12-1.432.O
463        084A  0-84A-y12-1.433.O
464        084A  0-84A-y12-1.434.I
465        084A  0-84A-y12-1.435.I
466        084X  0-84X-y12-1.452.I
467        084X  0-84X-y12-1.453.I
468        084X  0-84X-y12-1.454.O
469        084X  0-84X-y12-1.455.O

[470 rows x 2 columns]

In [11]:
gb_gjs = gps_JPID_StopID.groupby(['route_short','JourneyPatternID'])
gb_gjs_c = gb_gjs.count()
gb_gjs_c = gb_gjs_c.reset_index(['route_short','JourneyPatternID'])
gb_gjs_c.drop(['StopID', 'Count', 'Direction'], axis=1, inplace=True)
gb_gjs_c

route_short JourneyPatternID
0          0001         00010001
1          0001         00010002
2          0001         00010003
3          0001         00011001
4          0001         00011002
5          0004         00040001
6          0004         00040002
7          0004         00041001
8          0004         00041002
9          0007         00070001
10         0007         00070002
11         0007         00070003
12         0007         00070004
13         0007         00070005
14         0007         00071001
15         0007         00071002
16         0007         00071003
17         0007         00071004
18         0008         00080001
19         0008         00081001
20         0008         00081002
21         0009         00090001
22         0009         00090002
23         0009         00091001
24         0009         00091002
25         0011         00110001
26         0011         00110002
27         0011         00111001
28         0011         00111002
29         0013         00130001
..          ...              ...
453        067X         067X0003
454        067X         067X1001
455        067X         067X1002
456        068A         068A0001
457        068A         068A1001
458        069X         069X0001
459        069X         069X1001
460        0747         07470001
461        0747         07471001
462        076A         076A0001
463        076A         076A1001
464        077A         077A0001
465        077A         077A1001
466        079A         079A0001
467        079A         079A1001
468        083A         083A0001
469        083A         083A0002
470        083A         083A1001
471        083A         083A1002
472        084A         084A0001
473        084A         084A0002
474        084A         084A1001
475        084A         084A1002
476        084X         084X0001
477        084X         084X0002
478        084X         084X1001
479        084X         084X1002
480        86-2             86-2
481        OL77         OL77X101
482        PP07         PP071001

[483 rows x 2 columns]

In [12]:
# Route exist both in GPS and google transit
inter = set(gb_tss_c.route_short.unique()) & set(gb_gjs_c.route_short.unique())

tss_l = []
gjs_l = []

for r in inter:
    tss_l.append(gb_tss_c[gb_tss_c.route_short == r])
    gjs_l.append(gb_gjs_c[gb_gjs_c.route_short == r])

#     
tss_l[0] = tss_l[0].reset_index()
gjs_l[0] = gjs_l[0].reset_index()
df = pd.concat([tss_l[0], gjs_l[0]], axis=1)


for i in range(1, len(tss_l)):
    tss_l[i] = tss_l[i].reset_index()
    gjs_l[i] = gjs_l[i].reset_index()
    temp = pd.concat([tss_l[i], gjs_l[i]], axis=1)
    df = pd.concat([df, temp], axis=0)


In [13]:
# Route not in GPS data
diff = set(gb_tss_c.route_short.unique()) - inter

df2 = pd.DataFrame(columns = df.columns)
for r in diff:
    x = gb_tss_c[gb_tss_c.route_short == r]
    x = x.reset_index()
    y = gb_gjs_c[gb_gjs_c.route_short == r]
    y = y.reset_index()
    temp = pd.concat([x, y], axis=1)
    df2 = pd.concat([df2, temp], axis=0)

# Route not in google transit data

diff = set(gb_gjs_c.route_short.unique()) - inter

#df2 = pd.DataFrame(columns = df.columns)
for r in diff:
    x = gb_tss_c[gb_tss_c.route_short == r]
    x = x.reset_index()
    y = gb_gjs_c[gb_gjs_c.route_short == r]
    y = y.reset_index()
    temp = pd.concat([x, y], axis=1)
    df2 = pd.concat([df2, temp], axis=0)


In [14]:
out = pd.concat([df,df2], axis=0)
out.drop(['index'], axis=1, inplace=True)
out.to_csv('compare_route_googletransit_vs_gps.csv', index=False)

### Step2. Mapping the possible trip_id to JourneyPatternID
In order to find the possible trip_id to JourneyPatternID, here transfer the sequence of stop_id in each trip_id to set. The same as StopID of each JourneyPatternID.
The reason to transfer to a set is because we can use the set intersection method to find out the common StopID between two sets. Example, JourneyPatternID 00010001 has stopID {1,2,3,4}, 00010002 has {1,3,5} in gps data and trip_id 0-1-y12-1.1.O has stop_id {1,2,3,4}. So we can get the length of intersection from 00010001 is 4 and 00010002 is 2. Which means 00010001 is more likely is trip_id 0-1-y12-1.1.O

0720 Revised
#### Remain only those route is both exist in googletransit and gps data
Due to the observation above, we found out there are few route not exist in googletransit. So, we will delete it at this moment.

In [15]:
# In diff set are those route not exist in googletransit
inter = set(gb_tss_c.route_short.unique()) & set(gb_gjs_c.route_short.unique())
diff = set(gb_gjs_c.route_short.unique()) - inter

res['Route'] = res['JourneyPatternID'].str[:4]
res = res[~res.Route.isin(list(diff))]

#### 1. Get StopID set of googletransit data

In [16]:
# Sort by shape_id and stopID
transit_shapeID_stopID.sort_values(['shape_id', 'StopID'], inplace=True)

# Store each StopID of each shape into set
shapeID_stopID_set = {}
shape_unique = transit_shapeID_stopID.shape_id.unique()

for s in shape_unique:
    temp = []
    for stopid in transit_shapeID_stopID[transit_shapeID_stopID.shape_id == s]['StopID']:
        temp.append(stopid)
        temp.sort()
    shapeID_stopID_set[s] = [set(temp), temp]

In [17]:
# Here transform to dataframe so can merge later

# Get the shape_id to its stopID list
shapeID_stopID_df = pd.DataFrame(shapeID_stopID_set, index=['StopID_set', 'StopID_list'])
shapeID_stopID_df = shapeID_stopID_df.T
shapeID_stopID_df = shapeID_stopID_df.reset_index()
shapeID_stopID_df = shapeID_stopID_df.rename(columns = {'index':'shape_id'})

In [18]:
# Separate shape_id to route
shapeID_stopID_df['route_short'] = shapeID_stopID_df['shape_id'].apply(lambda x: x[x.index('-')+1:][:x[x.index('-')+1:].index('-')])
shapeID_stopID_df['route_short'] = shapeID_stopID_df['route_short'].apply(lambda x: str(x).zfill(4))

In [19]:
# Now we get the shape_id to its StopID set
shapeID_stopID_df

shape_id                                         StopID_set  \
0        0-1-y12-1.1.O  {0231, 0374, 0044, 0375, 0230, 0390, 0378, 035...   
1        0-1-y12-1.2.O  {0374, 0375, 0390, 0378, 0354, 0380, 0340, 037...   
2        0-1-y12-1.3.O  {0231, 0044, 0230, 0048, 1641, 0119, 0227, 021...   
3        0-1-y12-1.4.I  {0017, 0014, 0220, 0223, 0019, 0382, 0636, 037...   
4        0-1-y12-1.5.I  {0382, 0636, 0371, 0395, 0391, 0637, 0387, 039...   
5     0-102-y12-1.85.O  {3572, 4381, 0905, 6015, 0913, 3620, 3618, 093...   
6     0-102-y12-1.86.I  {3708, 3090, 3606, 4465, 1073, 6055, 4381, 360...   
7     0-102-y12-1.87.I  {3606, 4465, 1073, 4381, 3608, 0949, 3598, 095...   
8     0-104-y12-1.88.O  {1774, 4597, 0662, 0220, 0223, 1650, 0225, 165...   
9     0-104-y12-1.89.I  {0231, 1747, 0242, 1749, 0230, 0243, 0233, 176...   
10     0-11-y12-1.14.O  {0855, 0037, 0857, 0441, 0044, 0747, 0449, 026...   
11     0-11-y12-1.15.O  {0037, 0044, 0261, 6122, 0048, 0039, 0042, 004...   
12     0-11-y12-1.16.I  {0017, 0014, 0878, 0887, 0019, 3028, 0449, 302...   
13     0-11-y12-1.17.I  {0878, 0887, 3028, 0449, 3025, 0786, 3181, 088...   
14    0-111-y12-1.90.O  {3239, 3070, 7472, 3227, 3224, 3069, 3225, 323...   
15    0-111-y12-1.91.I  {3219, 3220, 3047, 3202, 4981, 3214, 2037, 321...   
16    0-114-y12-1.92.O  {3085, 3184, 0449, 6007, 3181, 3491, 3532, 045...   
17    0-114-y12-1.93.I  {3085, 2833, 3166, 3173, 3175, 3168, 3176, 317...   
18    0-116-y12-1.94.O  {0435, 2837, 2838, 0760, 0756, 2954, 2007, 298...   
19    0-116-y12-1.95.I  {0792, 3184, 0768, 2855, 0771, 2856, 0449, 285...   
20    0-116-y12-1.96.I  {0792, 3184, 0768, 2855, 0771, 2856, 0449, 285...   
21    0-118-y12-1.97.I  {0792, 0768, 0771, 3487, 0786, 3486, 3482, 472...   
22   0-120-y12-1.100.I  {0284, 5147, 0818, 6071, 0831, 0263, 7028, 008...   
23   0-120-y12-1.101.I  {0284, 5147, 0408, 0818, 0263, 0831, 0402, 041...   
24    0-120-y12-1.98.O  {0826, 5144, 4485, 0825, 0056, 0827, 7027, 082...   
25    0-120-y12-1.99.O  {0826, 5144, 0491, 4485, 0825, 0056, 0827, 040...   
26   0-122-y12-1.102.O  {1362, 0832, 6094, 1365, 1418, 1700, 4525, 001...   
27   0-122-y12-1.103.O  {0832, 1700, 4525, 0016, 1701, 0263, 0074, 008...   
28   0-122-y12-1.104.I  {0059, 0014, 1383, 0060, 1348, 0820, 1435, 135...   
29   0-122-y12-1.105.I  {1383, 1348, 1435, 1357, 1425, 1385, 1353, 143...   
..                 ...                                                ...   
440   0-83-y12-1.401.O  {0315, 4372, 2448, 1614, 4788, 2491, 7424, 033...   
441   0-83-y12-1.402.O  {0315, 4372, 1614, 4788, 7424, 0336, 1479, 161...   
442   0-83-y12-1.403.I  {0133, 2467, 0151, 0178, 1071, 7428, 2493, 489...   
443   0-83-y12-1.404.I  {2467, 1071, 2493, 2492, 1357, 1353, 1072, 247...   
444   0-83-y12-1.405.I  {0133, 2467, 0151, 0178, 1071, 7428, 2493, 489...   
445   0-83-y12-1.406.I  {0133, 0151, 0178, 7428, 4890, 0197, 0153, 015...   
446  0-83A-y12-1.407.O  {0315, 2448, 0622, 1614, 0599, 2491, 0336, 147...   
447  0-83A-y12-1.408.O  {0315, 2448, 1614, 4788, 2491, 7424, 0336, 147...   
448  0-83A-y12-1.409.I  {0133, 0159, 2467, 0151, 1071, 7428, 2493, 489...   
449  0-83A-y12-1.410.I  {0133, 0159, 2467, 0151, 1071, 7428, 2493, 489...   
450   0-84-y12-1.425.O  {4652, 4235, 4226, 4227, 4656, 6019, 4224, 421...   
451   0-84-y12-1.426.O  {3128, 4258, 4256, 4226, 4251, 4248, 4229, 311...   
452   0-84-y12-1.427.O  {3128, 3139, 4214, 4219, 4250, 4956, 3124, 422...   
453   0-84-y12-1.428.O  {4214, 4250, 4219, 4956, 4228, 4257, 4258, 424...   
454   0-84-y12-1.429.I  {4283, 4270, 3162, 4303, 4269, 3154, 7271, 314...   
455   0-84-y12-1.430.I  {4283, 4270, 3162, 4303, 4269, 3154, 7271, 314...   
456   0-84-y12-1.431.I  {4262, 4266, 4283, 4295, 7296, 4260, 4149, 427...   
457  0-84A-y12-1.432.O  {3128, 3139, 3124, 3135, 3130, 3133, 3118, 504...   
458  0-84A-y12-1.433.O  {3128, 3139, 3124, 3135, 0425, 3130, 3133, 311...   
459  0-84A-y12-1.434.I  {4202, 4416, 3143, 3079, 3156, 4203, 3159, 315...   
460  0-84A

#### 2. StopID set of gps data

In [20]:
# The same operation here for gps data
# Sort by JourneyPatternID and stopID
gps_JPID_StopID.sort_values(['JourneyPatternID', 'StopID'], inplace=True)

# Store each StopID of each JourneyPatternID into list
JPID_stopID_set = {}
JPID_unique = gps_JPID_StopID.JourneyPatternID.unique()

for s in JPID_unique:
    temp = []
    for stopid in gps_JPID_StopID[gps_JPID_StopID.JourneyPatternID == s]['StopID']:
        temp.append(stopid)
        temp.sort()
    JPID_stopID_set[s] = [set(temp), temp]

In [21]:
# Get the JourneyPatternID to its stopID list
JPID_stopID_df = pd.DataFrame(JPID_stopID_set, index=['StopID_set', 'StopID_list'])
JPID_stopID_df = JPID_stopID_df.T
JPID_stopID_df = JPID_stopID_df.reset_index()
JPID_stopID_df = JPID_stopID_df.rename(columns = {'index':'JourneyPatternID'})
JPID_stopID_df

JourneyPatternID                                         StopID_set  \
0           00010001  {0231, 0374, 0044, 0375, 0230, 0390, 0378, 035...   
1           00010002  {0374, 0375, 0390, 0378, 0354, 0380, 0340, 037...   
2           00010003  {0231, 0048, 0265, 1641, 0049, 0229, 0046, 005...   
3           00011001  {0017, 0014, 0220, 0223, 0019, 0382, 0371, 039...   
4           00011002  {0382, 0371, 0395, 0391, 0387, 0393, 0397, 039...   
5           00040001  {0113, 0037, 0419, 7481, 7483, 0408, 2026, 032...   
6           00040002  {0113, 0037, 0324, 0190, 0188, 0112, 0184, 014...   
7           00041001  {0092, 0091, 0483, 6182, 0197, 0473, 0028, 048...   
8           00041002  {0483, 7330, 0473, 0491, 0281, 3162, 0400, 047...   
9           00070001  {3219, 0419, 2035, 2041, 2042, 3220, 0408, 204...   
10          00070002  {3219, 0419, 2035, 2041, 2042, 3220, 0408, 204...   
11          00070003  {3219, 2035, 2041, 2042, 3220, 2043, 3047, 320...   
12          00070004  {3219, 0419, 2035, 3220, 0408, 3047, 3202, 042...   
13          00070005  {3219, 0419, 2035, 3220, 0408, 3047, 3202, 042...   
14          00071001  {2031, 0008, 2041, 3239, 3070, 7472, 3073, 048...   
15          00071002  {2031, 0008, 2041, 3239, 3070, 7472, 3073, 048...   
16          00071003  {0008, 2041, 3239, 3070, 7472, 3073, 0483, 307...   
17          00071004  {0008, 2041, 3239, 3070, 7472, 3073, 0483, 307...   
18          00080001  {0419, 0408, 6100, 3054, 0425, 0418, 2046, 040...   
19          00081001  {2031, 0008, 3058, 0483, 3073, 0473, 0480, 048...   
20          00081002  {0008, 3073, 0483, 3079, 0473, 0480, 0481, 610...   
21          00090001  {2389, 0037, 2392, 4392, 4788, 2415, 6228, 239...   
22          00090002  {0037, 4788, 0136, 0169, 0137, 0190, 0263, 018...   
23          00091001  {0130, 2432, 1348, 4392, 0091, 0131, 2459, 489...   
24          00091002  {2432, 1348, 4392, 2458, 2459, 2461, 1357, 135...   
25          00110001  {0855, 0037, 0857, 0441, 0044, 0747, 0449, 026...   
26          00110002  {0036, 0048, 0037, 0049, 0046, 0044, 0050, 004...   
27          00111001  {0017, 0014, 0878, 0887, 7220, 0019, 3028, 044...   
28          00111002  {0878, 0887, 3028, 0449, 0786, 0884, 0903, 087...   
29          00130001  {0110, 0113, 0037, 1944, 1947, 2160, 4549, 455...   
..               ...                                                ...   
453         067X0003  {3906, 1445, 3892, 2721, 3912, 7186, 1444, 390...   
454         067X1001  {0315, 3938, 2242, 3946, 3935, 0336, 0747, 147...   
455         067X1002  {0315, 3938, 2242, 3946, 3935, 0336, 0747, 147...   
456         068A0001  {1362, 1365, 1288, 1369, 1360, 1284, 1285, 128...   
457         068A1001  {1383, 1348, 1357, 1353, 1385, 1381, 1376, 138...   
458         069X0001  {3435, 1445, 2109, 3441, 1956, 2637, 1444, 343...   
459         069X1001  {2176, 3461, 4564, 5192, 1358, 3454, 3447, 345...   
460         07470001  {4724, 7409, 2507, 7491, 7489, 7492, 7402, 749...   
461         07471001  {1445, 7468, 1935, 7465, 7399, 7397, 3669, 739...   
462         076A0001  {2242, 2241, 2158, 2619, 2655, 2174, 2126, 702...   
463         076A1001  {2116, 2118, 2687, 2205, 4799, 7026, 2762, 268...   
464         077A0001  {2314, 4931, 2315, 2099, 2575, 2536, 7460, 233...   
465         077A1001  {4960, 2566, 2357, 2380, 2427, 5099, 2360, 238...   
466         079A0001  {1445, 6023, 2650, 2721, 6026, 1444, 2644, 265...   
467         079A1001  {2700, 6025, 2704, 2708, 0326, 1477, 2696, 269...   
468         083A0001  {0315, 2448, 1614, 4788, 2491, 7424, 0336, 147...   
469         083A0002  {0315, 2448, 1614, 4788, 2491, 7424, 0336, 147...   
470         083A1001  {0133, 0159, 2467, 0151, 1071, 7428, 2493, 489...   
471         083A1002  {0133, 0159, 2467, 0151, 1071, 7428, 2493, 489...   
472         084A0001  {3128, 3139, 3124, 3135, 3130, 3133, 3118, 504...   
473         084A0002  {3128, 3139, 3124, 3135, 0425, 3133, 3118, 313...   
474         084A1001  {4202, 

In [22]:
# Get the route from JourneyPatternID
JPID_stopID_df['route_short'] = JPID_stopID_df['JourneyPatternID'].str[:4]

In [23]:
# Now we also have JourneyPatternID to its StopID list
JPID_stopID_df

JourneyPatternID                                         StopID_set  \
0           00010001  {0231, 0374, 0044, 0375, 0230, 0390, 0378, 035...   
1           00010002  {0374, 0375, 0390, 0378, 0354, 0380, 0340, 037...   
2           00010003  {0231, 0048, 0265, 1641, 0049, 0229, 0046, 005...   
3           00011001  {0017, 0014, 0220, 0223, 0019, 0382, 0371, 039...   
4           00011002  {0382, 0371, 0395, 0391, 0387, 0393, 0397, 039...   
5           00040001  {0113, 0037, 0419, 7481, 7483, 0408, 2026, 032...   
6           00040002  {0113, 0037, 0324, 0190, 0188, 0112, 0184, 014...   
7           00041001  {0092, 0091, 0483, 6182, 0197, 0473, 0028, 048...   
8           00041002  {0483, 7330, 0473, 0491, 0281, 3162, 0400, 047...   
9           00070001  {3219, 0419, 2035, 2041, 2042, 3220, 0408, 204...   
10          00070002  {3219, 0419, 2035, 2041, 2042, 3220, 0408, 204...   
11          00070003  {3219, 2035, 2041, 2042, 3220, 2043, 3047, 320...   
12          00070004  {3219, 0419, 2035, 3220, 0408, 3047, 3202, 042...   
13          00070005  {3219, 0419, 2035, 3220, 0408, 3047, 3202, 042...   
14          00071001  {2031, 0008, 2041, 3239, 3070, 7472, 3073, 048...   
15          00071002  {2031, 0008, 2041, 3239, 3070, 7472, 3073, 048...   
16          00071003  {0008, 2041, 3239, 3070, 7472, 3073, 0483, 307...   
17          00071004  {0008, 2041, 3239, 3070, 7472, 3073, 0483, 307...   
18          00080001  {0419, 0408, 6100, 3054, 0425, 0418, 2046, 040...   
19          00081001  {2031, 0008, 3058, 0483, 3073, 0473, 0480, 048...   
20          00081002  {0008, 3073, 0483, 3079, 0473, 0480, 0481, 610...   
21          00090001  {2389, 0037, 2392, 4392, 4788, 2415, 6228, 239...   
22          00090002  {0037, 4788, 0136, 0169, 0137, 0190, 0263, 018...   
23          00091001  {0130, 2432, 1348, 4392, 0091, 0131, 2459, 489...   
24          00091002  {2432, 1348, 4392, 2458, 2459, 2461, 1357, 135...   
25          00110001  {0855, 0037, 0857, 0441, 0044, 0747, 0449, 026...   
26          00110002  {0036, 0048, 0037, 0049, 0046, 0044, 0050, 004...   
27          00111001  {0017, 0014, 0878, 0887, 7220, 0019, 3028, 044...   
28          00111002  {0878, 0887, 3028, 0449, 0786, 0884, 0903, 087...   
29          00130001  {0110, 0113, 0037, 1944, 1947, 2160, 4549, 455...   
..               ...                                                ...   
453         067X0003  {3906, 1445, 3892, 2721, 3912, 7186, 1444, 390...   
454         067X1001  {0315, 3938, 2242, 3946, 3935, 0336, 0747, 147...   
455         067X1002  {0315, 3938, 2242, 3946, 3935, 0336, 0747, 147...   
456         068A0001  {1362, 1365, 1288, 1369, 1360, 1284, 1285, 128...   
457         068A1001  {1383, 1348, 1357, 1353, 1385, 1381, 1376, 138...   
458         069X0001  {3435, 1445, 2109, 3441, 1956, 2637, 1444, 343...   
459         069X1001  {2176, 3461, 4564, 5192, 1358, 3454, 3447, 345...   
460         07470001  {4724, 7409, 2507, 7491, 7489, 7492, 7402, 749...   
461         07471001  {1445, 7468, 1935, 7465, 7399, 7397, 3669, 739...   
462         076A0001  {2242, 2241, 2158, 2619, 2655, 2174, 2126, 702...   
463         076A1001  {2116, 2118, 2687, 2205, 4799, 7026, 2762, 268...   
464         077A0001  {2314, 4931, 2315, 2099, 2575, 2536, 7460, 233...   
465         077A1001  {4960, 2566, 2357, 2380, 2427, 5099, 2360, 238...   
466         079A0001  {1445, 6023, 2650, 2721, 6026, 1444, 2644, 265...   
467         079A1001  {2700, 6025, 2704, 2708, 0326, 1477, 2696, 269...   
468         083A0001  {0315, 2448, 1614, 4788, 2491, 7424, 0336, 147...   
469         083A0002  {0315, 2448, 1614, 4788, 2491, 7424, 0336, 147...   
470         083A1001  {0133, 0159, 2467, 0151, 1071, 7428, 2493, 489...   
471         083A1002  {0133, 0159, 2467, 0151, 1071, 7428, 2493, 489...   
472         084A0001  {3128, 3139, 3124, 3135, 3130, 3133, 3118, 504...   
473         084A0002  {3128, 3139, 3124, 3135, 0425, 3133, 3118, 313...   
474         084A1001  {4202, 

#### 3. Now merge and find the length of common StopID between two set

In [24]:
# Merge two datafram base on the route_short (cause for now we still don't know which JourneyPatternID is which trip_id)
merge = pd.merge(shapeID_stopID_df, JPID_stopID_df, on='route_short', how='inner')

In [26]:
# Find the common StopID of googletransport and gps
# (x[0] & x[1]) is for python set structure to find the common elements between two sets
merge['common_len'] = merge.loc[:,['StopID_set_x', 'StopID_set_y']].apply(lambda x: len(x[0] & x[1]), axis=1)
merge['shape_id_len'] = merge['StopID_list_x'].apply(lambda x: len(x))
merge['JourneyPatternID_len'] = merge['StopID_list_y'].apply(lambda x: len(x))

merge['JPID_len-shapid_len'] = merge['JourneyPatternID_len'] - merge['shape_id_len']
merge = merge[merge['JPID_len-shapid_len'] == 0]

In [27]:
# Get the maxmun common len in each shape_id
merge_max = merge[merge.groupby(['shape_id'])['common_len'].transform('max') == merge['common_len']]
merge_max

shape_id                                       StopID_set_x  \
0         0-1-y12-1.1.O  {0231, 0374, 0044, 0375, 0230, 0390, 0378, 035...   
6         0-1-y12-1.2.O  {0374, 0375, 0390, 0378, 0354, 0380, 0340, 037...   
25     0-102-y12-1.85.O  {3572, 4381, 0905, 6015, 0913, 3620, 3618, 093...   
29     0-102-y12-1.86.I  {3708, 3090, 3606, 4465, 1073, 6055, 4381, 360...   
34     0-104-y12-1.88.O  {1774, 4597, 0662, 0220, 0223, 1650, 0225, 165...   
37     0-104-y12-1.89.I  {0231, 1747, 0242, 1749, 0230, 0243, 0233, 176...   
38      0-11-y12-1.14.O  {0855, 0037, 0857, 0441, 0044, 0747, 0449, 026...   
48      0-11-y12-1.16.I  {0017, 0014, 0878, 0887, 0019, 3028, 0449, 302...   
58     0-114-y12-1.92.O  {3085, 3184, 0449, 6007, 3181, 3491, 3532, 045...   
61     0-114-y12-1.93.I  {3085, 2833, 3166, 3173, 3175, 3168, 3176, 317...   
74    0-120-y12-1.100.I  {0284, 5147, 0818, 6071, 0831, 0263, 7028, 008...   
79    0-120-y12-1.101.I  {0284, 5147, 0408, 0818, 0263, 0831, 0402, 041...   
80     0-120-y12-1.98.O  {0826, 5144, 4485, 0825, 0056, 0827, 7027, 082...   
85     0-120-y12-1.99.O  {0826, 5144, 0491, 4485, 0825, 0056, 0827, 040...   
88    0-122-y12-1.102.O  {1362, 0832, 6094, 1365, 1418, 1700, 4525, 001...   
98    0-122-y12-1.104.I  {0059, 0014, 1383, 0060, 1348, 0820, 1435, 135...   
104   0-123-y12-1.106.O  {4399, 1467, 1486, 1490, 4496, 1425, 6057, 218...   
114   0-123-y12-1.108.I  {1494, 1460, 1455, 0509, 1935, 6094, 1499, 149...   
182     0-13-y12-1.21.I  {0017, 0014, 0092, 0091, 2154, 0088, 0281, 008...   
201     0-13-y12-1.22.I  {0017, 0014, 0092, 0091, 2154, 0088, 0281, 008...   
218     0-13-y12-1.23.I  {0017, 0014, 0092, 0091, 2154, 0088, 0281, 723...   
239     0-13-y12-1.24.I  {0017, 0014, 0092, 0091, 2154, 0088, 0281, 723...   
444   0-130-y12-1.110.O  {0515, 1747, 1749, 1740, 1753, 1752, 1767, 175...   
447   0-130-y12-1.111.I  {1774, 0617, 1739, 1756, 1775, 1778, 1735, 172...   
452    0-14-y12-1.185.I  {0515, 0669, 0242, 0297, 1187, 2842, 0243, 107...   
454    0-14-y12-1.186.O  {0617, 1081, 2855, 0234, 1047, 1195, 2864, 299...   
457   0-140-y12-1.112.O  {0747, 0403, 6363, 1017, 0190, 7149, 0191, 107...   
471   0-140-y12-1.115.I  {0792, 0893, 1071, 6228, 0897, 0197, 7111, 627...   
474   0-142-y12-1.116.I  {3606, 0336, 0403, 0905, 3630, 6310, 3629, 101...   
475   0-142-y12-1.117.O  {0893, 1071, 1073, 6055, 0897, 7197, 1072, 089...   
...                 ...                                                ...   
2562    0-7-y12-1.173.O  {3219, 0419, 2035, 3220, 0408, 3047, 3202, 042...   
2574    0-7-y12-1.174.I  {0008, 2041, 3239, 3070, 7472, 3073, 0483, 307...   
2580    0-7-y12-1.175.I  {0008, 2041, 3239, 3070, 7472, 3073, 0483, 307...   
2585   0-70-y12-1.271.O  {1445, 4325, 1805, 4770, 0784, 1664, 3331, 078...   
2588   0-70-y12-1.272.I  {4770, 0747, 1479, 0313, 1699, 1700, 1845, 190...   
2589  0-747-y12-1.165.O  {4724, 7409, 2507, 0612, 7402, 7406, 2499, 366...   
2592  0-747-y12-1.166.I  {0620, 1445, 7407, 7404, 1935, 7405, 7397, 366...   
2629   0-76-y12-1.391.O  {2206, 2158, 2619, 2655, 2174, 2126, 2159, 500...   
2632   0-76-y12-1.392.I  {2116, 2199, 2118, 2687, 2205, 4799, 2762, 268...   
2636  0-76A-y12-1.394.I  {2116, 2118, 2687, 2205, 4799, 7026, 2762, 268...   
2637  0-77A-y12-1.235.O  {2314, 4931, 2315, 2099, 2575, 2536, 7460, 233...   
2640  0-77A-y12-1.236.I  {4960, 2566, 2357, 2380, 2427, 5099, 2360, 238...   
2644   0-79-y12-1.396.I  {2700, 2704, 2670, 2708, 0326, 1477, 2696, 269...   
2645  0-79A-y12-1.397.O  {1445, 6023, 2650, 2721, 6026, 1444, 2644, 265...   
2648  0-79A-y12-1.398.I  {2700, 6025, 2704, 2708, 0326, 1477, 2696, 269...   
2652   0-7B-y12-1.177.I  {0792, 0008, 4731, 0768, 3555, 7056, 0771, 324...   
2657   0-7D-y12-1.179.I  {0792, 0008, 2063, 0768, 2057, 3062, 0771, 206...   
2660   0-7D-y12-1.180.I  {0792, 0008, 2063, 0768, 2057, 3062, 0771, 206...   
2662    0-8-y12-1.181.O  {0419, 0408, 6100, 3054, 0425, 0418, 2046, 040...   
2666    0-8-y12-1.182.I  {2031, 0008,

In [28]:
# Now we get the JourneyPatternID to its possible shape_id
# Here could have duplicate JPID but it's the same StopID_list so doesn't matter map to which shape_id, they are the same
map_JPId_shapeID = merge_max.loc[:, ['route_short', 'JourneyPatternID', 'shape_id', 'StopID_list_x']]
map_JPId_shapeID

route_short JourneyPatternID           shape_id  \
0           0001         00010001      0-1-y12-1.1.O   
6           0001         00010002      0-1-y12-1.2.O   
25          0102         01020001   0-102-y12-1.85.O   
29          0102         01021001   0-102-y12-1.86.I   
34          0104         01040001   0-104-y12-1.88.O   
37          0104         01041001   0-104-y12-1.89.I   
38          0011         00110001    0-11-y12-1.14.O   
48          0011         00111001    0-11-y12-1.16.I   
58          0114         01140001   0-114-y12-1.92.O   
61          0114         01141001   0-114-y12-1.93.I   
74          0120         01201001  0-120-y12-1.100.I   
79          0120         01201002  0-120-y12-1.101.I   
80          0120         01200001   0-120-y12-1.98.O   
85          0120         01200002   0-120-y12-1.99.O   
88          0122         01220001  0-122-y12-1.102.O   
98          0122         01221001  0-122-y12-1.104.I   
104         0123         01230001  0-123-y12-1.106.O   
114         0123         01231001  0-123-y12-1.108.I   
182         0013         00131001    0-13-y12-1.21.I   
201         0013         00131002    0-13-y12-1.22.I   
218         0013         00131001    0-13-y12-1.23.I   
239         0013         00131004    0-13-y12-1.24.I   
444         0130         01300001  0-130-y12-1.110.O   
447         0130         01301001  0-130-y12-1.111.I   
452         0014         00141001   0-14-y12-1.185.I   
454         0014         00140001   0-14-y12-1.186.O   
457         0140         01400001  0-140-y12-1.112.O   
471         0140         01401001  0-140-y12-1.115.I   
474         0142         01421001  0-142-y12-1.116.I   
475         0142         01420001  0-142-y12-1.117.O   
...          ...              ...                ...   
2562        0007         00070005    0-7-y12-1.173.O   
2574        0007         00071003    0-7-y12-1.174.I   
2580        0007         00070005    0-7-y12-1.175.I   
2585        0070         00700001   0-70-y12-1.271.O   
2588        0070         00701001   0-70-y12-1.272.I   
2589        0747         07470001  0-747-y12-1.165.O   
2592        0747         07471001  0-747-y12-1.166.I   
2629        0076         00760001   0-76-y12-1.391.O   
2632        0076         00761001   0-76-y12-1.392.I   
2636        076A         076A1001  0-76A-y12-1.394.I   
2637        077A         077A0001  0-77A-y12-1.235.O   
2640        077A         077A1001  0-77A-y12-1.236.I   
2644        0079         00791001   0-79-y12-1.396.I   
2645        079A         079A0001  0-79A-y12-1.397.O   
2648        079A         079A1001  0-79A-y12-1.398.I   
2652        007B         007B1001   0-7B-y12-1.177.I   
2657        007D         007D1001   0-7D-y12-1.179.I   
2660        007D         007D1001   0-7D-y12-1.180.I   
2662        0008         00080001    0-8-y12-1.181.O   
2666        0008         00081001    0-8-y12-1.182.I   
2707        0083         00831001   0-83-y12-1.403.I   
2718        0083         00831006   0-83-y12-1.404.I   
2745        083A         083A1001  0-83A-y12-1.409.I   
2764        0084         00840003   0-84-y12-1.427.O   
2793        084A         084A0001  0-84A-y12-1.432.O   
2803        084A         084A1001  0-84A-y12-1.434.I   
2817        084X         084X0001  0-84X-y12-1.454.O   
2824        084X         084X1002  0-84X-y12-1.455.O   
2825        0009         00090001     0-9-y12-1.10.O   
2835        0009         00091001     0-9-y12-1.12.I   

                                          StopID_list_x  
0     [0044, 0045, 0046, 0047, 0048, 0049, 0050, 005...  
6     [0265, 0271, 0340, 0350, 0351, 0352, 0353, 035...  
25    [0733, 0788, 0815, 0816, 0905, 0913, 0914, 093...  
29    [0928, 0944, 0945, 0947, 0948, 0949, 0950, 095...  
34    [0218, 0219, 0220, 0221, 0222, 0223, 0224, 022...  
37    [0226, 0227, 0228, 0229, 0230, 0231, 0232, 023...  
38    [0035, 0036, 0037, 0038, 0039, 0040, 0041, 004...  
48    [0010, 0012, 0014, 0015, 0017, 0018, 0019, 002...  
58    [044

In [29]:
# Export to csv
map_JPId_shapeID.to_csv('map_JPID_to_shape_id.csv', index=False)

In [31]:
res.shape

(70434885, 13)

#### Step 3. Now delete those rows that StopID should not in that JourneyPatternID

In [32]:
# Base on the mapping made before we can know what StopID should on that routes
# Before we already get the shape_id to its StopID sequence, so we use this as basic
transit_shapeID_stopID
# Add JourneyPatternID column
map_JPID_to_stopID = pd.merge(transit_shapeID_stopID, map_JPId_shapeID, on='shape_id', how='inner')
map_JPID_to_stopID = map_JPID_to_stopID.loc[:,['JourneyPatternID', 'StopID']]
map_JPID_to_stopID

JourneyPatternID StopID
0             00010001   0044
1             00010001   0045
2             00010001   0046
3             00010001   0047
4             00010001   0048
5             00010001   0049
6             00010001   0050
7             00010001   0051
8             00010001   0052
9             00010001   0119
10            00010001   0213
11            00010001   0214
12            00010001   0226
13            00010001   0227
14            00010001   0228
15            00010001   0229
16            00010001   0230
17            00010001   0231
18            00010001   0265
19            00010001   0271
20            00010001   0340
21            00010001   0350
22            00010001   0351
23            00010001   0352
24            00010001   0353
25            00010001   0354
26            00010001   0355
27            00010001   0356
28            00010001   0357
29            00010001   0372
...                ...    ...
11715         00091001   2432
11716         00091001   2433
11717         00091001   2434
11718         00091001   2435
11719         00091001   2436
11720         00091001   2437
11721         00091001   2438
11722         00091001   2439
11723         00091001   2440
11724         00091001   2441
11725         00091001   2442
11726         00091001   2443
11727         00091001   2444
11728         00091001   2446
11729         00091001   2459
11730         00091001   2460
11731         00091001   2461
11732         00091001   2462
11733         00091001   2463
11734         00091001   4369
11735         00091001   4376
11736         00091001   4377
11737         00091001   4392
11738         00091001   4479
11739         00091001   4680
11740         00091001   4890
11741         00091001   6219
11742         00091001   6229
11743         00091001   7131
11744         00091001   7458

[11745 rows x 2 columns]

#### 0720 revised
Split into 2 situation. For those JourneyPatternID certainly can map to certain trip_id, we delete those StopID should not on that JouurneyPatternID. Others, we still use percentage to judge if need to be remove.

#### First sub-set: JourneyPatternID can map to certain trip_id

In [33]:
# Get those JPID can map to trip_id
can_map = map_JPID_to_stopID['JourneyPatternID'].unique()
res_sub1 = res[res.JourneyPatternID.isin(can_map)]
res_sub1.shape

(56489727, 13)

In [34]:
# Now start to delete
# merge res and map_JPID_to_stopID

res_sub1 = pd.merge(res_sub1, map_JPID_to_stopID, on=['JourneyPatternID','StopID'], how='inner')
res_sub1

index           Timestamp JourneyPatternID  TimeFrame  \
0              135 2013-01-31 00:00:03         00071003 2013-01-30   
1              135 2013-01-31 00:00:03         00071003 2013-01-30   
2              136 2013-01-31 00:00:23         00071003 2013-01-30   
3              136 2013-01-31 00:00:23         00071003 2013-01-30   
4              137 2013-01-31 00:00:42         00071003 2013-01-30   
5              137 2013-01-31 00:00:42         00071003 2013-01-30   
6              138 2013-01-31 00:00:48         00071003 2013-01-30   
7              138 2013-01-31 00:00:48         00071003 2013-01-30   
8              139 2013-01-31 00:01:02         00071003 2013-01-30   
9              139 2013-01-31 00:01:02         00071003 2013-01-30   
10             140 2013-01-31 00:01:21         00071003 2013-01-30   
11             140 2013-01-31 00:01:21         00071003 2013-01-30   
12             141 2013-01-31 00:01:41         00071003 2013-01-30   
13             141 2013-01-31 00:01:41         00071003 2013-01-30   
14             142 2013-01-31 00:01:50         00071003 2013-01-30   
15             142 2013-01-31 00:01:50         00071003 2013-01-30   
16             143 2013-01-31 00:02:02         00071003 2013-01-30   
17             143 2013-01-31 00:02:02         00071003 2013-01-30   
18             144 2013-01-31 00:02:21         00071003 2013-01-30   
19             144 2013-01-31 00:02:21         00071003 2013-01-30   
20             145 2013-01-31 00:02:43         00071003 2013-01-30   
21             145 2013-01-31 00:02:43         00071003 2013-01-30   
22             146 2013-01-31 00:02:47         00071003 2013-01-30   
23             146 2013-01-31 00:02:47         00071003 2013-01-30   
24             147 2013-01-31 00:03:01         00071003 2013-01-30   
25             147 2013-01-31 00:03:01         00071003 2013-01-30   
26             148 2013-01-31 00:03:24         00071003 2013-01-30   
27             148 2013-01-31 00:03:24         00071003 2013-01-30   
28             149 2013-01-31 00:03:42         00071003 2013-01-30   
29             149 2013-01-31 00:03:42         00071003 2013-01-30   
...            ...                 ...              ...        ...   
60069454  62213316 2012-11-12 15:06:27         00410001 2012-11-12   
60069455  62213317 2012-11-12 15:06:43         00410001 2012-11-12   
60069456  62796559 2012-11-12 21:44:50         00410001 2012-11-12   
60069457  62796560 2012-11-12 21:45:04         00410001 2012-11-12   
60069458  62796561 2012-11-12 21:45:22         00410001 2012-11-12   
60069459  62796562 2012-11-12 21:45:43         00410001 2012-11-12   
60069460  63814554 2012-11-10 10:51:57         00410001 2012-11-10   
60069461  63814555 2012-11-10 10:52:11         00410001 2012-11-10   
60069462  63814556 2012-11-10 10:52:31         00410001 2012-11-10   
60069463  63814557 2012-11-10 10:52:52         00410001 2012-11-10   
60069464  64306444 2012-11-10 17:43:12         00410001 2012-11-10   
60069465  64306445 2012-11-10 17:43:24         00410001 2012-11-10   
60069466  64306446 2012-11-10 17:43:45         00410001 2012-11-10   
60069467  64306447 2012-11-10 17:44:05         00410001 2012-11-10   
60069468  64306448 2012-11-10 17:45:06         00410001 2012-11-10   
60069469  64323917 2012-11-10 18:06:13         00410001 2012-11-10   
60069470  64323918 2012-11-10 18:06:32         00410001 2012-11-10   
60069471  64323919 2012-11-10 18:06:55         00410001 2012-11-10   
60069472  66830529 2012-11-08 13:29:07         00410001 2012-11-08   
60069473  66830530 2012-11-08 13:29:26         00410001 2012-11-08   
60069474  66830531 2012-11-08 13:29:46         00410001 2012-11-08   
60069475  66830532 2012-11-08 13:30:07         00410001 2012-11-08   
60069476  68954455 2012-11-07 20:51:30         00410001 2012-11-07   
60069477  68954456 2012-11-07 20:51:32         00410001 2012-11-07   
60069478  68954457 2012-11-07 20:51:52         00410001 2012-11-07   
60069479  68954458 2012

In [37]:
res_sub1.shape

(60069484, 14)

In [ ]:
# Store as feather
#res_sub1 = res_sub1.reset_index()
res_sub1.to_feather('res_sub1.feather')

#### Second sub-set: If cannot map to any JournayPatternID
In version 3, use the outlier mean - 1.5*IQR to delete the outlier. In this version, for those JourneyPatternID not sure map to which trip_id we use "mean - 1.5*IQR" to delete the outlier.

In [ ]:
res_sub2 = res[~res.JourneyPatternID.isin(can_map)]
res_sub2.shape

In [ ]:
# Count how many times that StopID occurs in the certain JourneyPatternID

gb = res_sub2.groupby(['JourneyPatternID', 'StopID'])
idx = pd.DataFrame(gb['AtStop'].count())

In [ ]:
# Set index back to data

idx.reset_index(['JourneyPatternID', 'StopID'], inplace=True)

# Rename column

idx = idx.rename(columns = {'AtStop':'AtStopCount'})

In [ ]:
# In order to find the outlier, here calculate the mean and IQR of that JourneyPatternID + StopID

gb2 = idx.groupby(['JourneyPatternID'])
gb2count =  gb2['AtStopCount'].mean()
gb2q =  gb2['AtStopCount'].quantile(.75) - gb2['AtStopCount'].quantile(.25)
idx2 = pd.concat([pd.DataFrame(gb2count),pd.DataFrame(gb2q)], axis=1)

# In order to merge with idx, here set JourneyPatternID back to data

idx2.reset_index(['JourneyPatternID'], inplace=True)
idx2.columns = ['JourneyPatternID', 'AtStopCountMean', 'AtStopCountIQR']

In [ ]:
# Calculate mean - 2 * IQR

idx2['AtStopCountLowerBound'] = idx2['AtStopCountMean'] - 2* idx2['AtStopCountIQR']

In [ ]:
idx2

In [39]:
# Compare the time that StopID occurs of its JourneyPatternID with the lowerbound

merge = pd.merge(idx, idx2, on=['JourneyPatternID'], how='outer')

# merge will only keep those JourneyPatternID + StopID is outlier

merge = merge[merge['AtStopCount'] < merge['AtStopCountLowerBound']]

# Clean up..

merge.drop(['AtStopCount', 'AtStopCountMean', 'AtStopCountIQR'], axis=1, inplace=True)

In [40]:
# Merge with res, then those AtStopCountLowerBound is null should be kept

merge2 = pd.merge(res_sub2, merge, on=['JourneyPatternID', 'StopID'], how='outer',suffixes='OA')
res_sub2 = merge2[merge2['AtStopCountLowerBound'].isnull()]

# Clean up

res_sub2.drop('AtStopCountLowerBound', axis=1, inplace=True)
del merge, merge2

In [41]:
res_sub2

index           Timestamp JourneyPatternID  TimeFrame  \
0                0 2013-01-31 00:00:01         00131005 2013-01-30   
1              132 2013-01-30 23:59:04         00131005 2013-01-30   
2              133 2013-01-30 23:59:22         00131005 2013-01-30   
3              134 2013-01-30 23:59:43         00131005 2013-01-30   
4           526703 2013-01-29 23:58:10         00131005 2013-01-29   
5           526704 2013-01-29 23:58:28         00131005 2013-01-29   
6           526705 2013-01-29 23:58:49         00131005 2013-01-29   
7          1321942 2013-01-28 23:57:48         00131005 2013-01-28   
8          1321943 2013-01-28 23:58:06         00131005 2013-01-28   
9          1321944 2013-01-28 23:58:26         00131005 2013-01-28   
10         4160655 2013-01-27 00:05:03         00131005 2013-01-26   
11         4160656 2013-01-27 00:05:22         00131005 2013-01-26   
12         4160657 2013-01-27 00:05:43         00131005 2013-01-26   
13         4160658 2013-01-27 00:06:03         00131005 2013-01-26   
14         4160659 2013-01-27 00:06:48         00131005 2013-01-26   
15         4160660 2013-01-27 00:07:06         00131005 2013-01-26   
16         4836705 2013-01-25 23:57:08         00131005 2013-01-25   
17         4836706 2013-01-25 23:57:31         00131005 2013-01-25   
18         4836707 2013-01-25 23:57:49         00131005 2013-01-25   
19         4836708 2013-01-25 23:58:07         00131005 2013-01-25   
20         5842846 2013-01-25 00:00:03         00131005 2013-01-24   
21         5842977 2013-01-24 23:59:22         00131005 2013-01-24   
22         5842978 2013-01-24 23:59:44         00131005 2013-01-24   
23         7302859 2013-01-23 23:58:25         00131005 2013-01-23   
24         7302860 2013-01-23 23:58:45         00131005 2013-01-23   
25         8777145 2013-01-22 23:55:44         00131005 2013-01-22   
26         8777146 2013-01-22 23:56:07         00131005 2013-01-22   
27         8777147 2013-01-22 23:56:25         00131005 2013-01-22   
28        10149755 2013-01-21 23:56:49         00131005 2013-01-21   
29        10149756 2013-01-21 23:57:08         00131005 2013-01-21   
...            ...                 ...              ...        ...   
13945128  67063029 2012-11-08 16:39:37         00070003 2012-11-08   
13945129  68550004 2012-11-07 16:29:28         00070003 2012-11-07   
13945130  68550005 2012-11-07 16:29:48         00070003 2012-11-07   
13945131  68550006 2012-11-07 16:30:23         00070003 2012-11-07   
13945132  68550007 2012-11-07 16:30:45         00070003 2012-11-07   
13945133  68550008 2012-11-07 16:31:04         00070003 2012-11-07   
13945134  68550009 2012-11-07 16:31:24         00070003 2012-11-07   
13945135  68550010 2012-11-07 16:31:45         00070003 2012-11-07   
13945136  68550011 2012-11-07 16:32:03         00070003 2012-11-07   
13945137  68550012 2012-11-07 16:32:24         00070003 2012-11-07   
13945138  68550013 2012-11-07 16:32:46         00070003 2012-11-07   
13945139  67063030 2012-11-08 16:39:53         00070003 2012-11-08   
13945140  67063031 2012-11-08 16:40:14         00070003 2012-11-08   
13945141  67063032 2012-11-08 16:40:37         00070003 2012-11-08   
13945142  68550014 2012-11-07 16:33:03         00070003 2012-11-07   
13945143  68550015 2012-11-07 16:33:23         00070003 2012-11-07   
13945144  68550016 2012-11-07 16:33:44         00070003 2012-11-07   
13945145  67615527 2012-11-07 06:18:47         007D1002 2012-11-07   
13945146  67615528 2012-11-07 06:19:03         007D1002 2012-11-07   
13945147  67615529 2012-11-07 06:19:22         007D1002 2012-11-07   
13945148  67615530 2012-11-07 06:19:42         007D1002 2012-11-07   
13945149  67615531 2012-11-07 06:19:48         007D1002 2012-11-07   
13945150  67615532 2012-11-07 06:20:02         007D1002 2012-11-07   
13945151  67615533 2012-11-07 06:20:23         007D1002 2012-11-07   
13945152  67615534 2012-11-07 06:20:43         007D1002 2012-11-07   
13945153  67615535 2012

In [43]:
# Concat two files together

res = pd.concat([res_sub1, res_sub2], axis=0)
res

AtStop  AtStopCountLowerBound        Day  Delay JourneyPatternID  \
0           True                    NaN  Wednesday      0         00071003   
1           True                    NaN  Wednesday      0         00071003   
2           True                    NaN  Wednesday      0         00071003   
3           True                    NaN  Wednesday      0         00071003   
4           True                    NaN  Wednesday      0         00071003   
5           True                    NaN  Wednesday      0         00071003   
6           True                    NaN  Wednesday      0         00071003   
7           True                    NaN  Wednesday      0         00071003   
8           True                    NaN  Wednesday      0         00071003   
9           True                    NaN  Wednesday      0         00071003   
10          True                    NaN  Wednesday      0         00071003   
11          True                    NaN  Wednesday      0         00071003   
12          True                    NaN  Wednesday      0         00071003   
13          True                    NaN  Wednesday      0         00071003   
14          True                    NaN  Wednesday      0         00071003   
15          True                    NaN  Wednesday      0         00071003   
16          True                    NaN  Wednesday      0         00071003   
17          True                    NaN  Wednesday      0         00071003   
18          True                    NaN  Wednesday      0         00071003   
19          True                    NaN  Wednesday      0         00071003   
20          True                    NaN  Wednesday      0         00071003   
21          True                    NaN  Wednesday      0         00071003   
22          True                    NaN  Wednesday      0         00071003   
23          True                    NaN  Wednesday      0         00071003   
24          True                    NaN  Wednesday      0         00071003   
25          True                    NaN  Wednesday      0         00071003   
26          True                    NaN  Wednesday      0         00071003   
27          True                    NaN  Wednesday      0         00071003   
28          True                    NaN  Wednesday      0         00071003   
29          True                    NaN  Wednesday      0         00071003   
...          ...                    ...        ...    ...              ...   
13945128   False                    NaN   Thursday    695         00070003   
13945129   False                    NaN  Wednesday    141         00070003   
13945130   False                    NaN  Wednesday    163         00070003   
13945131   False                    NaN  Wednesday    190         00070003   
13945132   False                    NaN  Wednesday    211         00070003   
13945133   False                    NaN  Wednesday    231         00070003   
13945134   False                    NaN  Wednesday    250         00070003   
13945135   False                    NaN  Wednesday    250         00070003   
13945136   False                    NaN  Wednesday    266         00070003   
13945137   False                    NaN  Wednesday    283         00070003   
13945138   False                    NaN  Wednesday    283         00070003   
13945139   False                    NaN   Thursday    695         00070003   
13945140   False                    NaN   Thursday    713         00070003   
13945141   False                    NaN   Thursday    713         00070003   
13945142    True                    NaN  Wednesday    301         00070003   
13945143   False                    NaN  Wednesday    301         00070003   
13945144   False                    NaN  Wednesday    301         00070003   
13945145   False                    NaN  Wednesday    542         007D1002   
13945146   False                    NaN  Wednesday    542         007D1002   
13945147   False                

In [44]:
res.drop(['AtStopCountLowerBound', 'level_0', 'index'], axis=1, inplace=True)

AtStop        Day  Delay JourneyPatternID        Lat       Lon  \
0           True  Wednesday      0         00071003  53.317432 -6.230700   
1           True  Wednesday      0         00071003  53.317417 -6.230683   
2           True  Wednesday      0         00071003  53.317432 -6.230717   
3           True  Wednesday      0         00071003  53.317432 -6.230717   
4           True  Wednesday      0         00071003  53.317451 -6.230717   
5           True  Wednesday      0         00071003  53.317451 -6.230733   
6           True  Wednesday      0         00071003  53.317432 -6.230733   
7           True  Wednesday      0         00071003  53.317432 -6.230733   
8           True  Wednesday      0         00071003  53.317432 -6.230717   
9           True  Wednesday      0         00071003  53.317432 -6.230717   
10          True  Wednesday      0         00071003  53.317432 -6.230717   
11          True  Wednesday      0         00071003  53.317432 -6.230717   
12          True  Wednesday      0         00071003  53.317432 -6.230700   
13          True  Wednesday      0         00071003  53.317432 -6.230717   
14          True  Wednesday      0         00071003  53.317451 -6.230717   
15          True  Wednesday      0         00071003  53.317451 -6.230717   
16          True  Wednesday      0         00071003  53.317451 -6.230717   
17          True  Wednesday      0         00071003  53.317451 -6.230700   
18          True  Wednesday      0         00071003  53.317432 -6.230700   
19          True  Wednesday      0         00071003  53.317432 -6.230700   
20          True  Wednesday      0         00071003  53.317432 -6.230717   
21          True  Wednesday      0         00071003  53.317432 -6.230700   
22          True  Wednesday      0         00071003  53.317432 -6.230700   
23          True  Wednesday      0         00071003  53.317432 -6.230700   
24          True  Wednesday      0         00071003  53.317451 -6.230683   
25          True  Wednesday      0         00071003  53.317451 -6.230700   
26          True  Wednesday      0         00071003  53.317432 -6.230700   
27          True  Wednesday      0         00071003  53.317432 -6.230700   
28          True  Wednesday      0         00071003  53.317432 -6.230683   
29          True  Wednesday      0         00071003  53.317432 -6.230683   
...          ...        ...    ...              ...        ...       ...   
13945128   False   Thursday    695         00070003  53.281284 -6.138719   
13945129   False  Wednesday    141         00070003  53.285858 -6.145755   
13945130   False  Wednesday    163         00070003  53.285748 -6.145820   
13945131   False  Wednesday    190         00070003  53.284588 -6.146181   
13945132   False  Wednesday    211         00070003  53.284588 -6.146181   
13945133   False  Wednesday    231         00070003  53.284588 -6.146181   
13945134   False  Wednesday    250         00070003  53.284489 -6.146203   
13945135   False  Wednesday    250         00070003  53.283329 -6.145554   
13945136   False  Wednesday    266         00070003  53.282509 -6.142600   
13945137   False  Wednesday    283         00070003  53.282310 -6.141864   
13945138   False  Wednesday    283         00070003  53.281307 -6.138766   
13945139   False   Thursday    695         00070003  53.280308 -6.136205   
13945140   False   Thursday    713         00070003  53.280228 -6.136057   
13945141   False   Thursday    713         00070003  53.279705 -6.135986   
13945142    True  Wednesday    301         00070003  53.280720 -6.137417   
13945143   False  Wednesday    301         00070003  53.280304 -6.136192   
13945144   False  Wednesday    301         00070003  53.279991 -6.135425   
13945145   False  Wednesday    542         007D1002  53.340992 -6.258300   
13945146   False  Wednesday    542         007D1002  53.340992 -6.258300   
13945147   False  Wednesday    542         007D1002  53.340992 -6.258300   
13945148   False  Wednesday    542         007D1002  53.34

In [45]:
# Store as feather
res = res.reset_index()
res.to_feather('DBus_stage_three_clean_v1_step3.feather')

### Part III. Add Weather Information

In [3]:
res = pd.read_feather('DBus_stage_three_clean_v1_step3.feather')
res

index  AtStop        Day  Delay JourneyPatternID        Lat  \
0                0    True  Wednesday      0         00071003  53.317432   
1                1    True  Wednesday      0         00071003  53.317417   
2                2    True  Wednesday      0         00071003  53.317432   
3                3    True  Wednesday      0         00071003  53.317432   
4                4    True  Wednesday      0         00071003  53.317451   
5                5    True  Wednesday      0         00071003  53.317451   
6                6    True  Wednesday      0         00071003  53.317432   
7                7    True  Wednesday      0         00071003  53.317432   
8                8    True  Wednesday      0         00071003  53.317432   
9                9    True  Wednesday      0         00071003  53.317432   
10              10    True  Wednesday      0         00071003  53.317432   
11              11    True  Wednesday      0         00071003  53.317432   
12              12    True  Wednesday      0         00071003  53.317432   
13              13    True  Wednesday      0         00071003  53.317432   
14              14    True  Wednesday      0         00071003  53.317451   
15              15    True  Wednesday      0         00071003  53.317451   
16              16    True  Wednesday      0         00071003  53.317451   
17              17    True  Wednesday      0         00071003  53.317451   
18              18    True  Wednesday      0         00071003  53.317432   
19              19    True  Wednesday      0         00071003  53.317432   
20              20    True  Wednesday      0         00071003  53.317432   
21              21    True  Wednesday      0         00071003  53.317432   
22              22    True  Wednesday      0         00071003  53.317432   
23              23    True  Wednesday      0         00071003  53.317432   
24              24    True  Wednesday      0         00071003  53.317451   
25              25    True  Wednesday      0         00071003  53.317451   
26              26    True  Wednesday      0         00071003  53.317432   
27              27    True  Wednesday      0         00071003  53.317432   
28              28    True  Wednesday      0         00071003  53.317432   
29              29    True  Wednesday      0         00071003  53.317432   
...            ...     ...        ...    ...              ...        ...   
70422349  13945128   False   Thursday    695         00070003  53.281284   
70422350  13945129   False  Wednesday    141         00070003  53.285858   
70422351  13945130   False  Wednesday    163         00070003  53.285748   
70422352  13945131   False  Wednesday    190         00070003  53.284588   
70422353  13945132   False  Wednesday    211         00070003  53.284588   
70422354  13945133   False  Wednesday    231         00070003  53.284588   
70422355  13945134   False  Wednesday    250         00070003  53.284489   
70422356  13945135   False  Wednesday    250         00070003  53.283329   
70422357  13945136   False  Wednesday    266         00070003  53.282509   
70422358  13945137   False  Wednesday    283         00070003  53.282310   
70422359  13945138   False  Wednesday    283         00070003  53.281307   
70422360  13945139   False   Thursday    695         00070003  53.280308   
70422361  13945140   False   Thursday    713         00070003  53.280228   
70422362  13945141   False   Thursday    713         00070003  53.279705   
70422363  13945142    True  Wednesday    301         00070003  53.280720   
70422364  13945143   False  Wednesday    301         00070003  53.280304   
70422365  13945144   False  Wednesday    301         00070003  53.279991   
70422366  13945145   False  Wednesday    542         007D1002  53.340992   
70422367  13945146   False  Wednesday    542         007D1002  53.340992   
70422368  13945147   False  Wednesday    542         007D1002  53.340992   
70422369  13945148   False  Wednesday    542         007D1002

In [4]:
# Read weather data
ch = pd.read_csv('Casement_Hourly_Nov12Jan13_WindRainOnly.csv')
dh = pd.read_csv('DubAirport_Hourly_Nov12Jan13_WindRainOnly.csv')
ph = pd.read_csv('PhoenixPark_Hourly_Nov12Jan13_RainOnly.csv')

In [5]:
ph

date rain
0       6/11/2012 0:00    0
1       6/11/2012 1:00    0
2       6/11/2012 2:00    0
3       6/11/2012 3:00    0
4       6/11/2012 4:00    0
5       6/11/2012 5:00    0
6       6/11/2012 6:00    0
7       6/11/2012 7:00  0.1
8       6/11/2012 8:00    0
9       6/11/2012 9:00    0
10     6/11/2012 10:00    0
11     6/11/2012 11:00    0
12     6/11/2012 12:00    0
13     6/11/2012 13:00    0
14     6/11/2012 14:00    0
15     6/11/2012 15:00    0
16     6/11/2012 16:00    0
17     6/11/2012 17:00    0
18     6/11/2012 18:00    0
19     6/11/2012 19:00    0
20     6/11/2012 20:00    0
21     6/11/2012 21:00    0
22     6/11/2012 22:00    0
23     6/11/2012 23:00    0
24      7/11/2012 0:00    0
25      7/11/2012 1:00    0
26      7/11/2012 2:00    0
27      7/11/2012 3:00    0
28      7/11/2012 4:00    0
29      7/11/2012 5:00    0
...                ...  ...
1314  30/01/2013 18:00    0
1315  30/01/2013 19:00    0
1316  30/01/2013 20:00    0
1317  30/01/2013 21:00    0
1318  30/01/2013 22:00    0
1319  30/01/2013 23:00    0
1320  31/01/2013 00:00    0
1321  31/01/2013 01:00  2.3
1322  31/01/2013 02:00  0.8
1323  31/01/2013 03:00    0
1324  31/01/2013 04:00    0
1325  31/01/2013 05:00  0.1
1326  31/01/2013 06:00    0
1327  31/01/2013 07:00  0.2
1328  31/01/2013 08:00  0.3
1329  31/01/2013 09:00    0
1330  31/01/2013 10:00    0
1331  31/01/2013 11:00    0
1332  31/01/2013 12:00  0.1
1333  31/01/2013 13:00    0
1334  31/01/2013 14:00  0.1
1335  31/01/2013 15:00  2.5
1336  31/01/2013 16:00  0.5
1337  31/01/2013 17:00  0.2
1338  31/01/2013 18:00    0
1339  31/01/2013 19:00    0
1340  31/01/2013 20:00    0
1341  31/01/2013 21:00    0
1342  31/01/2013 22:00    0
1343  31/01/2013 23:00    0

[1344 rows x 2 columns]

In [6]:
dh.dtypes

date     object
rain    float64
wdsp      int64
dtype: object

In [7]:
from datetime import datetime
ch['DateTime'] = ch['date'].apply(lambda x: datetime.strptime(x, '%m/%d/%Y %H:%M'))
ch['DateTime'] = ch['DateTime'].apply(lambda x: datetime.strftime(x, '%Y-%m-%d %H'))

dh['DateTime'] = dh['date'].apply(lambda x: datetime.strptime(x, '%m/%d/%Y %H:%M'))
dh['DateTime'] = dh['DateTime'].apply(lambda x: datetime.strftime(x, '%Y-%m-%d %H'))


ph['DateTime'] = ph['date'].apply(lambda x: datetime.strptime(x, '%d/%m/%Y %H:%M'))
ph['DateTime'] = ph['DateTime'].apply(lambda x: datetime.strftime(x, '%Y-%m-%d %H'))

In [8]:
ch.drop('date', axis=1, inplace=True)
dh.drop('date', axis=1, inplace=True)
ph.drop('date', axis=1, inplace=True)

In [9]:
# Merge
merge = pd.merge(ch, dh, on='DateTime', how='outer')
merge = pd.merge(merge, ph, on='DateTime', how='outer')
merge

rain_x  wdsp_x       DateTime  rain_y  wdsp_y rain
0        0.0       8  2012-11-06 00     0.0      10    0
1        0.0       8  2012-11-06 01     0.0      10    0
2        0.0      10  2012-11-06 02     0.0      10    0
3        0.0      11  2012-11-06 03     0.0      13    0
4        0.0      10  2012-11-06 04     0.0      12    0
5        0.0      11  2012-11-06 05     0.0      12    0
6        0.0      11  2012-11-06 06     0.0      13    0
7        0.0      13  2012-11-06 07     0.0      16  0.1
8        0.0      13  2012-11-06 08     0.0      17    0
9        0.0      15  2012-11-06 09     0.0      16    0
10       0.0      15  2012-11-06 10     0.0      18    0
11       0.0      16  2012-11-06 11     0.0      19    0
12       0.0      16  2012-11-06 12     0.0      19    0
13       0.0      15  2012-11-06 13     0.0      19    0
14       0.0      14  2012-11-06 14     0.0      20    0
15       0.0      14  2012-11-06 15     0.0      19    0
16       0.0      13  2012-11-06 16     0.0      18    0
17       0.0      13  2012-11-06 17     0.0      15    0
18       0.0      13  2012-11-06 18     0.0      15    0
19       0.0      14  2012-11-06 19     0.0      18    0
20       0.0      13  2012-11-06 20     0.0      17    0
21       0.0      14  2012-11-06 21     0.0      14    0
22       0.0      14  2012-11-06 22     0.0      18    0
23       0.0      14  2012-11-06 23     0.0      17    0
24       0.0      15  2012-11-07 00     0.0      17    0
25       0.0      16  2012-11-07 01     0.0      16    0
26       0.0      15  2012-11-07 02     0.0      17    0
27       0.0      15  2012-11-07 03     0.0      17    0
28       0.0      14  2012-11-07 04     0.0      15    0
29       0.0      14  2012-11-07 05     0.0      16    0
...      ...     ...            ...     ...     ...  ...
1314     0.0      13  2013-01-30 18     0.0      17    0
1315     0.0      12  2013-01-30 19     0.0      16    0
1316     0.0      14  2013-01-30 20     0.0      18    0
1317     0.0      15  2013-01-30 21     0.0      17    0
1318     0.0      18  2013-01-30 22     0.0      10    0
1319     0.0      20  2013-01-30 23     0.0      11    0
1320     0.0      23  2013-01-31 00     0.0      16    0
1321     2.5      26  2013-01-31 01     1.2      21  2.3
1322     0.5      26  2013-01-31 02     1.3      25  0.8
1323     0.0      21  2013-01-31 03     0.0      20    0
1324     0.0      22  2013-01-31 04     0.0      22    0
1325     0.6      25  2013-01-31 05     0.1      24  0.1
1326     0.0      26  2013-01-31 06     0.2      23    0
1327     0.4      22  2013-01-31 07     0.0      20  0.2
1328     1.8      20  2013-01-31 08     0.0      24  0.3
1329     0.0      19  2013-01-31 09     0.0      18    0
1330     0.0      21  2013-01-31 10     0.0      22    0
1331     0.0      19  2013-01-31 11     0.0      22    0
1332     0.1      21  2013-01-31 12     0.0      23  0.1
1333     0.0      23  2013-01-31 13     0.0      24    0
1334     0.0      23  2013-01-31 14     0.1      24  0.1
1335     0.0      23  2013-01-31 15     2.2      24  2.5
1336     0.4      16  2013-01-31 16     0.1      21  0.5
1337     0.0      17  2013-01-31 17     0.7      18  0.2
1338     0.0      18  2013-01-31 18     0.0      15    0
1339     0.0      15  2013-01-31 19     0.0      16    0
1340     1.1      17  2013-01-31 20     0.0      18    0
1341     0.0      15  2013-01-31 21     0.0      17    0
1342     0.0      15  2013-01-31 22     0.0      18    0
1343     0.0      16  2013-01-31 23     0.0      18    0

[1344 rows x 6 columns]

In [10]:
merge.isnull().sum()

rain_x      0
wdsp_x      0
DateTime    0
rain_y      0
wdsp_y      0
rain        0
dtype: int64

In [11]:
# Convert to float
merge['rain'] = merge['rain'].convert_objects(convert_numeric=True)
merge.dtypes

/Users/EmmaChang/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  


rain_x      float64
wdsp_x        int64
DateTime     object
rain_y      float64
wdsp_y        int64
rain        float64
dtype: object

In [12]:
# Calculate average
merge['Rain_Avg'] = merge.loc[:,['rain_x', 'rain_y', 'rain']].mean(axis=1)
merge['Wind_Speed_Avg'] = merge.loc[:,['wdsp_x', 'wdsp_y']].mean(axis=1)
weather = merge.loc[:,['DateTime', 'Rain_Avg', 'Wind_Speed_Avg']]
weather

DateTime  Rain_Avg  Wind_Speed_Avg
0     2012-11-06 00  0.000000             9.0
1     2012-11-06 01  0.000000             9.0
2     2012-11-06 02  0.000000            10.0
3     2012-11-06 03  0.000000            12.0
4     2012-11-06 04  0.000000            11.0
5     2012-11-06 05  0.000000            11.5
6     2012-11-06 06  0.000000            12.0
7     2012-11-06 07  0.033333            14.5
8     2012-11-06 08  0.000000            15.0
9     2012-11-06 09  0.000000            15.5
10    2012-11-06 10  0.000000            16.5
11    2012-11-06 11  0.000000            17.5
12    2012-11-06 12  0.000000            17.5
13    2012-11-06 13  0.000000            17.0
14    2012-11-06 14  0.000000            17.0
15    2012-11-06 15  0.000000            16.5
16    2012-11-06 16  0.000000            15.5
17    2012-11-06 17  0.000000            14.0
18    2012-11-06 18  0.000000            14.0
19    2012-11-06 19  0.000000            16.0
20    2012-11-06 20  0.000000            15.0
21    2012-11-06 21  0.000000            14.0
22    2012-11-06 22  0.000000            16.0
23    2012-11-06 23  0.000000            15.5
24    2012-11-07 00  0.000000            16.0
25    2012-11-07 01  0.000000            16.0
26    2012-11-07 02  0.000000            16.0
27    2012-11-07 03  0.000000            16.0
28    2012-11-07 04  0.000000            14.5
29    2012-11-07 05  0.000000            15.0
...             ...       ...             ...
1314  2013-01-30 18  0.000000            15.0
1315  2013-01-30 19  0.000000            14.0
1316  2013-01-30 20  0.000000            16.0
1317  2013-01-30 21  0.000000            16.0
1318  2013-01-30 22  0.000000            14.0
1319  2013-01-30 23  0.000000            15.5
1320  2013-01-31 00  0.000000            19.5
1321  2013-01-31 01  2.000000            23.5
1322  2013-01-31 02  0.866667            25.5
1323  2013-01-31 03  0.000000            20.5
1324  2013-01-31 04  0.000000            22.0
1325  2013-01-31 05  0.266667            24.5
1326  2013-01-31 06  0.066667            24.5
1327  2013-01-31 07  0.200000            21.0
1328  2013-01-31 08  0.700000            22.0
1329  2013-01-31 09  0.000000            18.5
1330  2013-01-31 10  0.000000            21.5
1331  2013-01-31 11  0.000000            20.5
1332  2013-01-31 12  0.066667            22.0
1333  2013-01-31 13  0.000000            23.5
1334  2013-01-31 14  0.066667            23.5
1335  2013-01-31 15  1.566667            23.5
1336  2013-01-31 16  0.333333            18.5
1337  2013-01-31 17  0.300000            17.5
1338  2013-01-31 18  0.000000            16.5
1339  2013-01-31 19  0.000000            15.5
1340  2013-01-31 20  0.366667            17.5
1341  2013-01-31 21  0.000000            16.0
1342  2013-01-31 22  0.000000            16.5
1343  2013-01-31 23  0.000000            17.0

[1344 rows x 3 columns]

In [13]:
# Merge with gps data
res['DateTime'] = res['Timestamp'].apply(lambda x: datetime.strftime(x, '%Y-%m-%d %H'))
res_2 = pd.merge(res, weather, on='DateTime', how='left')
res_2

index  AtStop        Day  Delay JourneyPatternID        Lat  \
0                0    True  Wednesday      0         00071003  53.317432   
1                1    True  Wednesday      0         00071003  53.317417   
2                2    True  Wednesday      0         00071003  53.317432   
3                3    True  Wednesday      0         00071003  53.317432   
4                4    True  Wednesday      0         00071003  53.317451   
5                5    True  Wednesday      0         00071003  53.317451   
6                6    True  Wednesday      0         00071003  53.317432   
7                7    True  Wednesday      0         00071003  53.317432   
8                8    True  Wednesday      0         00071003  53.317432   
9                9    True  Wednesday      0         00071003  53.317432   
10              10    True  Wednesday      0         00071003  53.317432   
11              11    True  Wednesday      0         00071003  53.317432   
12              12    True  Wednesday      0         00071003  53.317432   
13              13    True  Wednesday      0         00071003  53.317432   
14              14    True  Wednesday      0         00071003  53.317451   
15              15    True  Wednesday      0         00071003  53.317451   
16              16    True  Wednesday      0         00071003  53.317451   
17              17    True  Wednesday      0         00071003  53.317451   
18              18    True  Wednesday      0         00071003  53.317432   
19              19    True  Wednesday      0         00071003  53.317432   
20              20    True  Wednesday      0         00071003  53.317432   
21              21    True  Wednesday      0         00071003  53.317432   
22              22    True  Wednesday      0         00071003  53.317432   
23              23    True  Wednesday      0         00071003  53.317432   
24              24    True  Wednesday      0         00071003  53.317451   
25              25    True  Wednesday      0         00071003  53.317451   
26              26    True  Wednesday      0         00071003  53.317432   
27              27    True  Wednesday      0         00071003  53.317432   
28              28    True  Wednesday      0         00071003  53.317432   
29              29    True  Wednesday      0         00071003  53.317432   
...            ...     ...        ...    ...              ...        ...   
70422349  13945128   False   Thursday    695         00070003  53.281284   
70422350  13945129   False  Wednesday    141         00070003  53.285858   
70422351  13945130   False  Wednesday    163         00070003  53.285748   
70422352  13945131   False  Wednesday    190         00070003  53.284588   
70422353  13945132   False  Wednesday    211         00070003  53.284588   
70422354  13945133   False  Wednesday    231         00070003  53.284588   
70422355  13945134   False  Wednesday    250         00070003  53.284489   
70422356  13945135   False  Wednesday    250         00070003  53.283329   
70422357  13945136   False  Wednesday    266         00070003  53.282509   
70422358  13945137   False  Wednesday    283         00070003  53.282310   
70422359  13945138   False  Wednesday    283         00070003  53.281307   
70422360  13945139   False   Thursday    695         00070003  53.280308   
70422361  13945140   False   Thursday    713         00070003  53.280228   
70422362  13945141   False   Thursday    713         00070003  53.279705   
70422363  13945142    True  Wednesday    301         00070003  53.280720   
70422364  13945143   False  Wednesday    301         00070003  53.280304   
70422365  13945144   False  Wednesday    301         00070003  53.279991   
70422366  13945145   False  Wednesday    542         007D1002  53.340992   
70422367  13945146   False  Wednesday    542         007D1002  53.340992   
70422368  13945147   False  Wednesday    542         007D1002  53.340992   
70422369  13945148   False  Wednesday    542         007D1002

In [14]:
res_2 = res_2.reset_index()
res_2.to_feather('res_2.feather')

### Part III. Add School Holiday

In [15]:
res_2 = pd.read_feather('res_2.feather')
res_2.shape

(70422379, 17)

In [22]:
school_holiday = ['2013-01-01','2013-01-02', '2013-01-03','2013-01-04','2013-01-05','2013-01-06']
school_holiday_days = ['Saturday','Sunday']
res_2['SchoolHoliday'] = np.zeros(res_2.shape[0], dtype=bool)

level_0     index  AtStop        Day  Delay JourneyPatternID  \
0                0         0    True  Wednesday      0         00071003   
1                1         1    True  Wednesday      0         00071003   
2                2         2    True  Wednesday      0         00071003   
3                3         3    True  Wednesday      0         00071003   
4                4         4    True  Wednesday      0         00071003   
5                5         5    True  Wednesday      0         00071003   
6                6         6    True  Wednesday      0         00071003   
7                7         7    True  Wednesday      0         00071003   
8                8         8    True  Wednesday      0         00071003   
9                9         9    True  Wednesday      0         00071003   
10              10        10    True  Wednesday      0         00071003   
11              11        11    True  Wednesday      0         00071003   
12              12        12    True  Wednesday      0         00071003   
13              13        13    True  Wednesday      0         00071003   
14              14        14    True  Wednesday      0         00071003   
15              15        15    True  Wednesday      0         00071003   
16              16        16    True  Wednesday      0         00071003   
17              17        17    True  Wednesday      0         00071003   
18              18        18    True  Wednesday      0         00071003   
19              19        19    True  Wednesday      0         00071003   
20              20        20    True  Wednesday      0         00071003   
21              21        21    True  Wednesday      0         00071003   
22              22        22    True  Wednesday      0         00071003   
23              23        23    True  Wednesday      0         00071003   
24              24        24    True  Wednesday      0         00071003   
25              25        25    True  Wednesday      0         00071003   
26              26        26    True  Wednesday      0         00071003   
27              27        27    True  Wednesday      0         00071003   
28              28        28    True  Wednesday      0         00071003   
29              29        29    True  Wednesday      0         00071003   
...            ...       ...     ...        ...    ...              ...   
70422349  70422349  13945128   False   Thursday    695         00070003   
70422350  70422350  13945129   False  Wednesday    141         00070003   
70422351  70422351  13945130   False  Wednesday    163         00070003   
70422352  70422352  13945131   False  Wednesday    190         00070003   
70422353  70422353  13945132   False  Wednesday    211         00070003   
70422354  70422354  13945133   False  Wednesday    231         00070003   
70422355  70422355  13945134   False  Wednesday    250         00070003   
70422356  70422356  13945135   False  Wednesday    250         00070003   
70422357  70422357  13945136   False  Wednesday    266         00070003   
70422358  70422358  13945137   False  Wednesday    283         00070003   
70422359  70422359  13945138   False  Wednesday    283         00070003   
70422360  70422360  13945139   False   Thursday    695         00070003   
70422361  70422361  13945140   False   Thursday    713         00070003   
70422362  70422362  13945141   False   Thursday    713         00070003   
70422363  70422363  13945142    True  Wednesday    301         00070003   
70422364  70422364  13945143   False  Wednesday    301         00070003   
70422365  70422365  13945144   False  Wednesday    301         00070003   
70422366  70422366  13945145   False  Wednesday    542         007D1002   
70422367  70422367  13945146   False  Wednesday    542         007D1002   
70422368  70422368  13945147   False  Wednesday    542         007D1002   
70422369  70422369  13945148   False  Wednesday    542         007D1002   
70422370  70422370  13945149   False

In [28]:
res_2.loc[res_2.TimeFrame.isin(school_holiday), 'SchoolHoliday'] = True
res_2.loc[res_2.Day.isin(school_holiday_days), 'SchoolHoliday'] = True

In [29]:
res_2.to_feather('DBus_stage_two_clean_v4.feather')

In [27]:
res_2.shape

(70422379, 18)